src/login/org.freedesktop.login1.conf

```diff
@@ -350,6 +350,10 @@
                        send_interface="org.freedesktop.login1.Session"
                        send_member="SetDisplay"/>
 
+                <allow send_destination="org.freedesktop.login1"
+                       send_interface="org.freedesktop.login1.Session"
+                       send_member="SetTTY"/>
```

Comment:
i think it would be safest if before issuing ttyname_r() we check that `fcntl(F_GETFL) & O_ACCMODE == O_RDWR` holds.

And for paranoia maybe also check that O_PATH is off. (though ttyname_r() in glibc first thing does is TCGETS which should fail on O_PATH fds)

Reply:
> hmm, i wonder about the security implications of this. the only restriction the current PR makes is that you must have taken control of the session, and that's something you can do when your uid matches the UID of the user owning the session.

But if there are two users with the same UID, you have a security problem anyways since the one can always access all files of the other?

> Maybe instead of accepting a tty path we should accept an fd to a tty. That way we know for sure the client has the privs to open the tty and that should probably enough for us.

If my uid machtes the UID of the user owning the session, then I can also open his TTY device file. Because this is owned by the same UID.

@poettering I understand your concerns, but I don't see a difference with your suggestion. Maybe we can restrict this by only allowing to set a TTY if none is set? So the typical sshd case, where the tty is not yet known during authentication, but shortly afterwards? Leads of course still to the problem to verify, that the tty string is a valid device.

Reply:
> But if there are two users with the same UID, you have a security problem anyways since the one can always access all files of the other?

people who have two users with the same UID are doing things wrong. And of course there's no isolation, hence it's fine if they can trample on each other's sessions.

> If my uid machtes the UID of the user owning the session, then I can also open his TTY device file. Because this is owned by the same UID.

Exactly. What I am concerned about though is that with the current patch people might report *any* tty to logind and it would take it happilly, regardless who it is owned by: the session's own user or some other user.

Reply:
I changed to code to use a fd and added checks that the device is RW and no O_PATH set.
We can still relax the checks later if really necessary.